File: java/0077_combine.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> comb = new ArrayList<Integer>();
        backtrack(res,comb,n,1,k);
        return res;
    }
    public void backtrack(List<List<Integer>> res, List<Integer> comb, int n,int pos,int k)
    {
        if(comb.size()==k)
        {
            res.add(new ArrayList<Integer>(comb));
            return;
        }
        for(int i=pos;i<=n;i++)
        {
            comb.add(i);
            backtrack(res,comb,n,i+1,k);
            comb.remove(comb.size()-1);
        }
    }
}

File: cpp/0077_combine.cpp
// 普通回溯
// 1.回溯范围：所有1到n的组合
// 2.剪枝条件：无
// 3.终止：达到要求长度k
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        vector<int> comb;
        backtrack(res,comb,n,1,k);
        return res;
    }
    void backtrack(vector<vector<int>> &res,vector<int>&comb,int n,int pos,int k)
    {
        if(comb.size()==k)
        {
            res.push_back(comb);
            return;
        }
        for(int i=pos;i<=n;i++)
        {
            comb.push_back(i);
            backtrack(res,comb,n,i+1,k);
            comb.pop_back();
        }
    }
};
